cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -0,0 +1,141 @@
+ */
+import cpp
+import semmle.code.cpp.ir.IR
+
+/**
+ * A node in a data flow graph.
+ *
+ * A node can be either an expression, a parameter, or an uninitialized local
+ * variable. Such nodes are created with `DataFlow::exprNode`,
+ * `DataFlow::parameterNode`, and `DataFlow::uninitializedNode` respectively.
+*/
+class Node extends Instruction {
+  /**
+   * INTERNAL: Do not use. Alternative name for `getFunction`.
+   */
+  Function getEnclosingCallable() {
+    result = this.getFunction()
+  }
+
+  /** Gets the type of this node. */
+  Type getType() {
+    result = this.getResultType()
+  }
+
+  /** Gets the expression corresponding to this node, if any. */
+  Expr asExpr() { result = this.getUnconvertedResultExpression() }
+
+  /** Gets the parameter corresponding to this node, if any. */
+  Parameter asParameter() { result = this.(InitializeParameterInstruction).getParameter() }
+
+  /**
+   * Gets the uninitialized local variable corresponding to this node, if
+   * any.
+   */
+  LocalVariable asUninitialized() {
+    result = this.(UninitializedInstruction).getLocalVariable()
+  }
+
+  /**
+   * Gets an upper bound on the type of this node.
+   */
+  Type getTypeBound() { result = getType() }
+}
+
+/**
+ * An expression, viewed as a node in a data flow graph.
+ */
+class ExprNode extends Node {
+  ExprNode() { getAST() instanceof Expr }
+  Expr getExpr() { result = getAST() }
+}
+
+/**
+ * The value of a parameter at function entry, viewed as a node in a data
+ * flow graph.
+ */
+class ParameterNode extends Node, InitializeParameterInstruction {
+  /**
+   * Holds if this node is the parameter of `c` at the specified (zero-based)
+   * position. The implicit `this` parameter is considered to have index `-1`.
+   */
+  predicate isParameterOf(Function f, int i) {
+    f.getParameter(i) = getParameter()
+  }
+}
+
+/**
+ * The value of an uninitialized local variable, viewed as a node in a data
+ * flow graph.
+ */
+class UninitializedNode extends Node, UninitializedInstruction {
+}
+
+/**
+ * A node associated with an object after an operation that might have
+ * changed its state.
+ *
+ * This can be either the argument to a callable after the callable returns
+ * (which might have mutated the argument), or the qualifier of a field after
+ * an update to the field.
+ *
+ * Nodes corresponding to AST elements, for example `ExprNode`, usually refer
+ * to the value before the update with the exception of `ClassInstanceExpr`,
+ * which represents the value after the constructor has run.
+ */
+abstract class PostUpdateNode extends Node {
+  /**
+   * Gets the node before the state update.
+   */
+  abstract Node getPreUpdateNode();
+}
+
+class StoreDestinationAsPostUpdateNode extends PostUpdateNode {
+  StoreInstruction si;
+  StoreDestinationAsPostUpdateNode() {
+    this = si.getDestinationAddress()
+  }
+  
+  override Node getPreUpdateNode() {
+    result = si.getDestinationAddress()
```

Comment:
Isn't this the same as `result = this`? If so, what's the need for the `si` field?

Reply:
That's definitely wrong. That class isn't used yet. `PostUpdateNode` is part of the Java field flow implementation, which we're not using yet. I've just removed `StoreDestinationAsPostUpdateNode` for now.